{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bc37a725",
   "metadata": {},
   "source": [
    "### Kinds of Plots\n",
    "\n",
    "- The \"toddler\" plot--all curves from all patients colored by case/control with data points for each sample (DONE)\n",
    "- One curve (one color) for case/control with data points for each sample (DONE)\n",
    "- One point for case/control at each location (average across sites)\n",
    "\n",
    "### Tracks\n",
    "\n",
    "- Promoter region / etc\n",
    "- CpG islands\n",
    "- GC content\n",
    "- Expression?\n",
    "\n",
    "### Zoom\n",
    "Window at boundaries; then +/- 100; 1000; 10,000\n",
    "\n",
    "### Color\n",
    "https://cran.r-project.org/web/packages/ggsci/vignettes/ggsci.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "cca95bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DSS)\n",
    "library(Gviz)\n",
    "library(tidyverse)\n",
    "library(data.table)\n",
    "library(ggsci)\n",
    "library(parallel)\n",
    "library(scales)\n",
    "library(wiscR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "655ba6f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m2\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m4\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (2): chr, name\n",
      "\u001b[32mdbl\u001b[39m (2): start, end\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# user inputs\n",
    "CHR <- \"chr18\"\n",
    "GEN <- \"hg38\"\n",
    "FS <- 18 # fontsize\n",
    "\n",
    "#TODO: csv with chr,start,stop,Name \n",
    "# Then filter by chr matching our CHR\n",
    "# supp.df <- read_csv()\n",
    "idir <- paste0(\"./result-\", CHR, \"/\")\n",
    "\n",
    "roi.df <- read_csv(\"regions-of-interest.csv\")\n",
    "M <- fread(file.path(idir, \"M.csv\"))\n",
    "Cov <- fread(file.path(idir, \"Cov.csv\"))\n",
    "Meth <- fread(file.path(idir, \"methylation.csv\"))\n",
    "load(file.path(idir, \"test-values.RData\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7a0fe729",
   "metadata": {},
   "outputs": [],
   "source": [
    "NCOL <- 7\n",
    "colors <- pal_nejm(\"default\")(NCOL)\n",
    "#show_col(colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a8816435",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m83\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m18\u001b[39m\n",
      "\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m  (6): cohort, machine, pool, group, batch, sex\n",
      "\u001b[32mdbl\u001b[39m (12): sample, CD8T, CD4T, NK, Bcell, Mono, Gran, charlson_score, bmi, wa...\n",
      "\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use \u001b[30m\u001b[47m\u001b[30m\u001b[47m`spec()`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set \u001b[30m\u001b[47m\u001b[30m\u001b[47m`show_col_types = FALSE`\u001b[47m\u001b[30m\u001b[49m\u001b[39m to quiet this message.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " [1] 131 154 144 120 167 185 188 109 126 166 142 146 111 145 159 164 105 100 129\n",
      "[20] 112 125 147 115 104 170 110 108 107 136 161 143 148 174 158 116 132 134 186\n",
      "[39] 140 127 114 130 122 133 156 139 101 119 123 137 118 138 141 169 106 157 165\n",
      "[58] 163 135 113 168\n"
     ]
    }
   ],
   "source": [
    "df <- read_csv(\"../../data/meta/phenos-cleaned.csv\")\n",
    "#TODO: make parameters from command line...\n",
    "valid_samples <- intersect(colnames(Meth)[-1], df$sample)\n",
    "print(valid_samples)\n",
    "\n",
    "filt.df <- df %>% filter(sample %in% valid_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "954f410c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dmrs <- callDMR(test.cohort, p.threshold=0.01, dis.merge = 2500, pct.sig = 0.5, minCG = 3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "917e63af",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "`stat_bin()` using `bins = 30`. Pick better value with `binwidth`.\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAC/VBMVEUAAAABAQECAgIDAwME\nBAQFBQUGBgYHBwcICAgJCQkKCgoLCwsMDAwNDQ0ODg4PDw8QEBARERESEhITExMUFBQVFRUW\nFhYXFxcYGBgZGRkaGhobGxscHBwdHR0eHh4fHx8gICAhISEiIiIjIyMkJCQlJSUmJiYnJyco\nKCgpKSkqKiorKyssLCwtLS0uLi4vLy8wMDAxMTEyMjIzMzM0NDQ1NTU2NjY3Nzc4ODg5OTk6\nOjo7Ozs8PDw9PT0+Pj4/Pz9AQEBBQUFCQkJDQ0NERERFRUVGRkZHR0dISEhJSUlKSkpLS0tM\nTExNTU1OTk5PT09QUFBRUVFSUlJTU1NUVFRVVVVWVlZXV1dYWFhZWVlaWlpbW1tcXFxdXV1e\nXl5fX19gYGBhYWFiYmJjY2NkZGRlZWVmZmZnZ2doaGhpaWlqampra2tsbGxtbW1ubm5vb29w\ncHBxcXFycnJzc3N0dHR1dXV2dnZ3d3d4eHh5eXl6enp7e3t8fHx9fX1+fn5/f3+BgYGCgoKD\ng4OEhISFhYWGhoaHh4eIiIiJiYmKioqLi4uMjIyNjY2Ojo6Pj4+QkJCRkZGSkpKTk5OUlJSV\nlZWWlpaXl5eYmJiZmZmampqbm5ucnJydnZ2enp6fn5+goKChoaGioqKjo6OkpKSlpaWmpqan\np6eoqKipqamqqqqrq6usrKytra2urq6vr6+wsLCxsbGysrKzs7O0tLS1tbW2tra3t7e4uLi5\nubm6urq7u7u8vLy9vb2+vr6/v7/AwMDBwcHCwsLDw8PExMTFxcXGxsbHx8fIyMjJycnKysrL\ny8vMzMzNzc3Ozs7Pz8/Q0NDR0dHS0tLT09PU1NTV1dXW1tbX19fY2NjZ2dna2trb29vc3Nzd\n3d3e3t7f39/g4ODh4eHi4uLj4+Pk5OTl5eXm5ubn5+fo6Ojp6enq6urr6+vs7Ozt7e3u7u7v\n7+/w8PDx8fHy8vLz8/P09PT19fX29vb39/f4+Pj5+fn6+vr7+/v8/Pz9/f3+/v7////6NCC7\nAAAACXBIWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3dDZxVdYH/8avgU5qtpaVZ61bbs7pt\n5f9v61aWm7bWDKgoiQ+FJppJLiqhQKtUPgW5EirqooZI+ECIASJKGBg+xG5I+ZSihKbCwDAM\nw+PA/F7L/Z17z8zce879nnPu78ycy/18Xq96Oefxd+fcN3Mfzj03lyMiIiKiXaijRhBR4o4q\nQLrEEFHiLgESUfUBichBQCJyEJCIHFSvkLZsS3f77c2b0t2B2ZDy9tc3p7yDjTvS3f6W5i3p\n7mD7xi4/1CukzVvT3f62prZ0d2DWp7z9dWtS3kHb9nS3v7lpc7o72N71EAMpnYAkA9KuEJBU\nQFIByQBJByQVkAyQdEBSAckASQckFZAMkHRAUgHJAEkHJBWQDJB0QFIByQBJByQVkAyQdEBS\nAckASQckFZAMkHRAUgHJAEkHJBWQDJB0QFIByQBJByQVkAyQdEBSAckASQckFZAMkHRAUgHJ\nAEkHJBWQDJB0QFIByQBJByQVkAyQdEBSAckASQckFZAMkHRAUgHJAEkHJBWQDJB0QFIByQBJ\nByQVkAyQdEBSAckASQck1S4M6fSAgpcEkgpIKiAZIOmApAKSAZIOSCogGSDpgKQCkgGSDkgq\nIBkg6YCkApIBkg5IKiAZIOmApAKSAZIOSCogGSDpgKQCkgGSDkgqIBkg6YCkApIBkg5IKiAZ\nIOmApAKSAZIOSCogGSDpgKQCkgGSDkgqIBkg6YCkApIBkg5IKiAZIOmApAKSAZIOSCogGSDp\ngKQCkgGSDkgqIBkg6YCkApIBkg5IKiAZIOmApAKSAZIOSCogGSDpgKQCkgGSDkgqIBkg6YCk\nApIBkg5IKiAZIOmApAKSAZIOSCogGSDpgKQCkgGSDkgqIBkg6YCkApIBkg5IKiAZIOmApAKS\nAZIOSCogGSDpgKQCkgGSDkgqIBkg6YCkApIBkg5IKiAZIOmApAKSAZIOSCogGSDpgKQCkgGS\nDkgqIBkg6YCkApIBkg5IKiAZIOmApAKSAZIOSCogGSDpgKQCkgGSDkgqIBkg6YCkApIBkg5I\nKiAZIOmApAKSAZIOSCogGSDpgKQCkgGSDkgqIBkg6YCkApIBkg5IKiAZIOmApAKSAZIOSCog\nGSDpgKSqZUjbt1QsCFLwkhvaKm+p2jY2rU93B1vWpbz95jUp72D9pnS3v6FpQ7o72OQdYu+f\n5BqDtLmpYkGQKq9BVGXN9p5ZY5DaN1UsCFLwkq0bKm+p2jY0taS7g03NKW9/7ZqUd9DSlu72\nW5vWp7uDNu8Qew8gawySiOdI7uI5kqqWnyOJgOQuIKmAZICkA5IKSAZIOiCpgGSApAOSCkgG\nSDogqYBkgKQDkgpIBkg6IKmAZICkA5IKSAZIOiCpgGSApAOSCkgGSDogqYBkgKQDkgpIBkg6\nIKmAZICkA5IKSAZIOiCpgGSApAOSCkgGSDogqYBkgKQDkgpIBkg6IKmAZICkA5IKSAZIOiCp\ngGSApAOSCkgGSDogqYBkgKQDkgpIBkg6IKmAZICkA5IKSAZIOiCpgGSApAOSCkgGSDogqYBk\ngKQDkgpIBkg6IKmAZICkA5IKSAZIOiCpgGSApAOSCkgGSDogqYBkgKQDkgpIBkg6IKmAZICk\nA5IKSAZIOiCpgGSApAOSCkgGSDogqYBkgKQDkgpIBkg6IKmAZICkA5IKSAZIOiCpgGSApAOS\nCkgGSDogqYBkgKQDkgpIBkg6IKmAZICkA5IKSAZIOiCpgGSApAOSCkgGSDogqYBkgKQDkgpI\nBkg6IKmAZICkA5IKSAZIOiCpgGSApAOSCkgGSDogqYBkgKQDkgpIBkg6IKmAZICkA5IKSAZI\nOiCpgGSApAOSCkgGSDogqYBkgKQDkgpIBkg6IKmAZICkA5IKSAZIOiCpgGSApAOSCkgGSDog\nqYBkgKQDkgpIBkg6IKmAZICkA5IKSAZIOiCpgGSApAOSCkgGSDogqYBkgKQDkgpIBkg6IKmA\nZICkA5IKSAZIOiCpgGSApAOSCkgGSDogqTINaf3SedOn/Xr+81sSrQ0kdwFJ1WuQNt/UkG9m\n6AKr7r6woVD/y+ck+DUAyV1AUvUWpBeHNFSEtHFCv4auDZrbEXcXQHIXkFS9A2n7lCKTEEgr\nz20o7eq493YguQtIql6B9MYwn0cwpJWnlzlqaBjVHm8vQHIXkFS9AKlj1ikNlSFtOMeb2W/0\nfY89PW9qkd2N8fYDJHcBSdXzkNb+p8fiO+GQfuYtMbGl8PPKUd6ExbF2BCR3AUnV45AWDbIm\nTnl4Tiik5+ycxvmdU3bcYCedG+vBHZDcBSRVT0Oa6v1tGbrShEMaaedM6TZM7+Hd7Di7ApK7\ngKTqaUjXew/atplwSCvsjCHdf7PLG/MTL4yzKyC5C0iq3oB0xjP5/wqFdLud8WjJ1DF26gsx\ndgUkdwFJ1QuQRq21/xUKyb6FdFrpaUFP2cUnxdgVkNwFJFWPQ+p/f+EUhTBIr9rp15ZO3mpf\nMz8/xq6A5C4gqXoa0uSXiv8VBmmWnT63bPpoO31d9F0ByV1AUvXe2d9hkMbZ6a+WTZ8c960k\nILkLSKrsQRqan9y//Pe60C4/NfoOgOQuIKmyB+lU++J3+fLL7fLjou8ASO4CkipzkNbZyaPK\nl19vZ1wafQdAcheQVJmD5L1oN7Z8+Y7+9iyh6DsAkruApMocpGft5FsDVjgrP+PU6DsAkruA\npMocpCfs5MkBK3ifqd0WeQdAcheQVJmDNN9Ovi9ghYvtnOh3SSC5C0iqzEF6xE6eHrCCdwJ4\nS8Cc4IDkLiCpMgfJm/xgwArD7Zymzgk72isWBCl4yY2bKm+p2jY3taa7g/aWlLffvCblHbRu\nTXf7G5s2pruDrd4h9v49yAKk2aGQLiuF1L6pYkGQgpds3VB5S9W2oakl3R1sak55+2vXpLyD\nlrZ0t9/a1JruDtq8Q+z93csCJB7aJYmHdqL6e2gX/mLDUF5sCA1IovqDtNhO/mXACt+1c6L/\nwoHkLiCpMgfpT3byLQEr2KumDIy+AyC5C0iqzEH6q51c9rm+nSNtDDmbNSwguQtIqsxBarOT\nLy9ffrWdMTL6DoDkLiCpMgfJewQ3uHz5ZXb5CdF3ACR3AUmVPUj2fdfG8pvtvcE0I/oOgOQu\nIKmyB2minf5c2fQbQ6aHBiR3AUmVPUgLQt6RvSA/+aQYd3kguQtIquxBWtsY+KLCW3bxgE/O\nhgYkdwFJlT1I5pL89P6l19263y4+K8YOgOQuIKkyCGlG0ElC28+zvJpj7ABI7gKSKoOQWk/O\nzxjU/b43N+x92vCA5C4gqTIIyUywc37edVKT912YL8bZAZDcBSRVFiE1eV+NeU/nlBbvzO+f\nxNoBkNwFJFUWIZnpdlbDNW95P+5YPNj+PHB1rB0AyV1AUvUwpJMbwpvmL9UxujBp+JS5C2dO\nPNv7oXFhvH0ByV1AUmUSkmkbGjA/6FOzlQKSu4CkyiYk0zaydO5J5V/0IgKSu4Ckyigk0zF9\nYLeZw1+LvS8guQtIqqxCMmbDjGGNhTnfGvdsgn0ByV1AUvXeq3YR2rB07gPTHly0oiPR2kBy\nF5BUmYZUXUByF5BUQDJA0gFJBSQDJB2QVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUjiGt\nmJEv5V9KxIDkLiCpIkM6P9/TanPz7Pp/dDe+KgKSu4CkigzJzrhbbW61XSzoO8l7PiC5C0gq\nx5DMO/KLneNseNUEJHcBSeUa0mH5xf6/s+FVE5DcBSSVa0iH5hf7gLPhVROQ3AUklWNIrX3y\ni+3nbnxVBCR3AUnlGNLNdrG+7sZXRUByF5BUTiHtmLKPXWwflyNMHJDcBSSVgPToiGJ2xjdH\nhPfD7ze8t7D+B9Mdc8SA5C4gqQSkK3NJOiHdMUcMSO4CkioVSGPTHXPEgOQuIKnSgLT/2nTH\nHDEguQtIqjQg3ZLukKMGJHcBSSUgLbviy++IyajPT9MdceSA5C4gqfTL3+1Lxp9+WGRGux//\nZLoDjh6Q3AUkVcT3kd6838448oRK9Ttr+NTX0x1unIDkLiCpXJ9rl6GA5C4gqYBkgKQDkgpI\nBkg6IKmAZICkA5IqMqQv55uX7mDcBiR3AUnF5bgMkHRAUgHJAEkHJBWQDJB0QFIByQBJByQV\nkAyQdEBSAckASQckVWxIO/4y+/Zrr7yiwmfOR4xId8hRA5K7gKSKB6n5phPeGeEU8HSHHDUg\nuQtIqjiQ2i7fN4IiIJUFJFk9QXrp49EYAak0IMnqCNKq6J/uS3fIUQOSu4Ckigxp+9GRHQGp\nJCDJ6gfS3dEdAakkIMnqBlLHp7pA2f+jn/vCMRVKd8hRA5K7gKSKCmlREdEhQx9J+5fqKCC5\nC0iqqJAKF7jb/Ueb0h2Pw4DkLiCpokJq8Obclu5onAYkdwFJFRXSJ+2Mb6Q7GLcByV1AUkWF\n9Hd2xoJ0B+M2ILkLSKqokPbIT99zS7qDcRuQ3AUkVVRI9rthj0x3LI4DkruApIr10O4r6Y7F\ncUByF5BUUSHZE+36pTsWxwHJXUBSRYV0LH+RkgckWd1AsrM+me5YHAckdwFJFRXSjPz0fWrn\ntAYDJJcBSRUV0hb7asOD6Q7GbUByF5BUkT9GMSw/Y0C6g3EbkNwFJFVkSE375+c8ke5onAYk\ndwFJFf2j5nfl53x6XbrDcRmQ3AUkVYyLn3w/P+sLad8l3AUkdwFJFQNSxw/y8z7xVLoDcheQ\n3AUkVawLRE7KX9euzzn/m+6QXAUkdwFJFe9Kq6/8u53/uUvufWFVyne+6gOSu4Ckinzxk1+O\nv3rkD84946Cc3557hZXukKMGJHcBSRUVUnsuRukOOWpAcheQVEAyQNIBSQUkAyQdkFRAMkDS\nAUkFJAMkHZBUQDJA0gFJFRXSjqtvvH3KvTNmzZ7zsC7dIUcNSO4CkoovYzZA0gFJBSQDJB2Q\nVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUQDJA0gFJBSQDJB2QVEAyQNIBSQUkAyQdkFSR\nIb0Qoef/vPSZ38+fNS0bn0IHkruApIoMKc5Jq9m4+B2Q3AUkFZAMkHRAUtUypB3tFQuCFLzk\nxk2Vt1Rtm5ta091Be0vK229ek/IOWremu/2NTRvT3cFW7xBvr0VIm5sqFgSp8hpEVdZci5C2\nbahYEKTgJVvWV95Sta1vWpfuDjY0p7z9tWtS3sG61nS339LUku4OWr1DvLEWIYl4juQuniOp\nnD5Hes8Xj2sYNOTSq7akO+aIAcldQFJFhvRWWG+++OScaxoPzK+3+/ez8bfIC0juApLK1ZkN\n7fcdm1/z0AWuBlZ9QHIXkFQOTxG6L/9Xafcb3YzLQUByF5BULs+1e/Nj+ZWvczIuBwHJXUBS\nOT1pdeUH8n+TMnI1LiA5DEgqt2d/P5hf+30Z+XZMILkLSCrHH6P4t/zqV1U/LBcByV1AUjmG\ndE9+9YNT/qVEDEjuApLKMaSWPfLrz6t+XA4CkruApHL9CdmP5te/ouphuQhI7gKSyjWkL+bX\nP6HqYbkISO4Ckso1pMb8+h+uelguApK7gKRyDekr9uTVqoflIiC5C0gq15A+k19/j6qH5SIg\nuQtIKseQtu2dX3/f6sflICC5C0gqx5AW2/U/WP24HAQkdwFJ5RjSQLv+kdWPy0FAcheQVG4h\nzdzdrt/gYGDVByR3AUnlEtKO2/by1r/WydCqDUjuApLKGaSOl8b+U2H13O9dja6qgOQuIKki\nQxpYodP6f/Xw/XJ+/9Ce7pgjBiR3AUmVxpVWb0p3yFEDkruApEoB0kdTHnLUgOQuIKncQzrg\npXRHHDkguQtIKueQDlyU7oCjByR3AUnlGNJup72R7nhjBCR3AUnlFNLHLnsh3dHGCkjuApIq\nMqQzKnXm4At/dPO81ekONW5AcheQVHwZswGSDkgqIBkg6YCkApIBkg5IKiAZIOmApAKSAZIO\nSCogGSDpgKQCkgGSDkgqIBkg6YCkSgppx8pnZt9zxy0T7/zVI0sz9kZsMSC5C0iqJJBeuuXs\nw/fuemrQu4+55MG0D2X8gOQuIKliQ3p7zCcDz7Pbq/9DHekONW5AcheQVDEhrfreHuHnrH7i\ngXTHGjMguQtIqniQpr4rnFG+E95Kd7SxApK7gKSKA6njB5UZ7eyQp9IdbpyA5C4gqeJAOlc6\nyuXemY1LceUDkruApIoB6WcRHOVy73413QFHD0juApIqOqQ/dXmZ4R1Hn/eTu2b+7sn/WbJ4\nwfTbRw/61O6d8z6bja9iBpLLgKSKDunL/vOgyxaWXQFy/QOn++8s3ZDaYOMFJHcBSRUZ0u8L\n8957+7bgLa0Zvqe3xAEb0hho/IDkLiCpIkMa5M36WoXf6NIPe8v8l+tBJgtI7gKSKiqkLe+0\nc06oeFnvtz9iF/pnx2NMGJDcBSRVVEiP2RnvW1t5a8v62sVecznCxAHJXUBSRYX0Uzvj52pz\n37OL3eVwgMkDkruApIoK6VQ74021uT/axb7ncIDJA5K7gKSKCumo/PRPys11HJhf7jh346si\nILkLSKqokN6fn/5tvb2T8sv9o7PhVROQ3AUkVVRI9kW74Xp7F9t3kpwNr5qA5C4gqaJCsi/H\nRfiS5TH55fZyNrxqApK7gKSKBenHenvj8sv1cTa8agKSu4CkivXQ7mK9vSvyy73T2fCqCUju\nApIqKqRD89NP1dsbnF/u/c6GV01AcheQVFEhHZ2ffrDe3sfyyx3tbHjVBCR3AUkVFZJ3zuoy\ntbnX7GKDHA4weUByF5BUUSHZFxFyw9TmRkU7k6hHApK7gKSKCulJO2Pvlypv7eV32MWecTrE\npAHJXUBSRYW042A754iKR2ztp+xCh2bjSpFAcheQVJE/2Fe4FNdnVoZv68XCNVgvdT3IZAHJ\nXUBSRYb0cuH6Ju+6cUvwllqv2sdbou9f0xho/IDkLiCpol/85KzCzNzBly8pe+zW/vj39i/O\n/35qg40XkNwFJFV0SKsOyPkddOLI/35kyfI3Vq16/eVn5ky87Lj9O+e9vyXdEUcOSO4CkirG\nBSIf2i0XoT0zc6lVILkLSKo4lyweF8FR33vTHW+MgOQuIKlifRvFTX2Uo3fOTHe4cQKSu4Ck\nive1Lgs/WNnR58Ubtj0akNwFJFXMLxprG7VvOKNDJ+5Id7DxApK7gKSK/dWXzdd8PFDRbsfc\nkfLdMW5AcheQVEm+jHnZzxoP7o7oQ9+amJF3YbsEJHcBSZUEUr6WZx689dqrRo4ac/3tv/nf\njakNr5qA5C4gqZJCqoGA5C4gqYBkgKQDkgpIBkg6IKmAZICkA5LKEaRMvYFUCEjuApIqPqT1\nv75watnEPx5w/FUvux5blQHJXUBSxYX08tD9crkryibbr3P50kPux1dFQHIXkFTxIG253J62\nel7ZDO97kXJfd3Ku3aSGsGJdVQVI7gKSKhaklwoXZTipbE4BUm7fOQ7GNB5IcQOSKFOQniue\nGPSlsllFSLm+d1Y/pmuAFDcgibIE6fWDilo+VTbPh5Tr83jVYxoNpLgBSZQhSB3H+VjeWzaz\nE1Lu4LeqHdMwIMUNSKIMQZrgU9nv+LKZrbcOfk9xdsi9NXrn5c2MrnYrQHIYkFSRIW19f2Hm\nsQ+2B25py53v9RbY/cUqxzQoD+maKjcCJJcBSRUZ0h3erHdMCd/Wmn/xljmzyjH1z0MaX+VG\ngOQyIKkiQ/q8nbP34kobaz3CLrRXdVe222yfDk2qahv5gOQuIKmiQvqbd1G7WypvbYl3maHy\nU4ji1GQhTatqG/mA5C4gqaJCut/O+LA6ObW/Xay6ixavsJBmVbWNfEByF5BUUSH9yM4Yozb3\nG7vYF6sa0nMW0oKqtpEPSO4CkioqpLPtjAVqc+vsI8C/r2pIz1hIT1e1jXxAcheQVFEhfcPO\neFNuz75Ivl9VQ1pgIf25qm3kA5K7gKSKCumrdoYezOH5xfpUNaRZFtKKqraRD0juApIqKqSv\n2xnb5Pbsl1/uU9WQpllITVVtIx+Q3AUkVVRIp9oZ+tdpTxA/sKoh3WEhbdr5X6sevfnaUZeO\nvP7up5NcOQ9I7gKSKiqki+wM+QpAs13sE1UN6Rd5R/1M+2NDO89WPena+A/1gOQuIKmiQhpv\nZ9yoNvcru9g3qxrStXk5pz97XvcTvxuvi3u+BJDcBSRVVEgLor1BdKxd7IdVDcl+HKmx/DMU\n574abztAcheQVFEhbfG+snxR5a3N8Nav7vPmoR9HGvCXWNsBkruApIp80uq/2zkfr3jEnn+3\nXWjvDVUNaUgRznmTlq7dtnXVsruKj/IGx3p0ByR3AUkVGZJ3sl3umHXh21r8Pm+Zs6sb0hmF\nR3ILOwc5d6A37apuC3ZULghS8JKbtohNVdnWprZ0d9CxPuXtr1uT8g7a2tPd/k5I6e6gvXCI\nJaT2Q715h4U9bnvr0r7eErv9oTpI9uNIDWO7/ZVYea4n6YWuE7dtqFgQpOAlW9ZX3lLc7ZW2\nvmldtAUT15zy9tc2pbyDda3pbr+lqSXdHbR6h3ijhGQmF2bmjp5U/mbpxnln7VmcX+VHzTsW\n5ivFuGJAQ9wPoLt+aBd9e6Xx0E5WPw/tjPlKUUpu9yO+ffXU3z772pur31z+pyfvH3vRv+zh\nz8u952+pDHSG9ydpVfQ1gOQuIKliQHrzA7kI7f6bdAa65UwLaXb0NYDkLiCp4lwg8o9/FwHS\nbWmN1J7w0DAu+gpAcheQVLEuWfz8hxWjve9ObaQLLaRLo68AJHcBSRXvIvqrT6vs6Igl6Y30\nZe+tpOgrAMldQFLF/VqXeYeHMzporP6YRfJWWUiDoq8AJHcBSRX/i8Z+e/IegYyOmpjuveVN\nC+ms6CsAyV1AUiX56svWmRce3reroX3/3zm3rkx1mDt7yUI6P/oKQHIXkFRJv0N2ywuPTL7x\nmjE/HX/X9MeWd6Q2vC49YSGNjL4CkNwFJFU2v9W8pbl82kQLSVyhsmtAcheQVJmD9PoNI4ec\n0jCqfMYFFtL86FsCkruApMocpBb7kb5+ZacCLfE+JxvwpyosILkLSKrMQSp8ru8XJVO3X2In\nXxljQ0ByF5BU2YM01/vTU3L6953eOatxLhsJJHcBSZU9SFu9D/YN6HqSREfB0Yg4GwKSu4Ck\nyh6kwiWLGxon+M+Tlg33Jg3Ul0zuEpDcBSRVBiGZ6wqXaOg3/K45Cx9/6KbvFn5ufCrWZoDk\nLiCpsgipfUzgRYROjvHSdz4guQtIqixCMu23BlzW7uy43/EMJHcBSZVJSMa8cFkJo9Onxb78\nN5DcBSRVRiEZ88rNF/h/lvqPmJngKvpAcheQVJmFtLPWpY/Puu++mQv/kuxjTkByF5BUWYZU\nZUByF5BUQDJA0gFJBSQDJB2QVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUQDJA0gFJBSQD\nJB2QVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUQDJA0gFJBSQDJB2QVEAyQNIBSQUkAyQd\nkFRAMkDSAUkFJAMkHZBUQDJA0gFJBSQDJB2QVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBU\nQDJA0gFJBSQDJB2QVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUQDJA0gFJBSQDJB2QVEAy\nQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUQDJA0gFJBSQDJB2QVEAyQNIBSQUkAyQdkFRAMkDS\nAUkFJAMkHZBUQDJA0gFJBSQDJB2QVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUQDJA0gFJ\nBSQDJB2QVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUQDJA0gFJBSQDJB2QVEAyQNIBSQUk\nAyQdkFRAMkDSAUkFJAMkHZBUQDJA0gFJBSQDJB2QVEAyQNIBSQUksytDSj7m7gFJBSQDJB2Q\nVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUQDJA0gFJBSQDJB2QVEAyQNIBSQUkAyQdkFRA\nMkDSAUkFJAMkHZBUQDJA0gFJBSQDJB2QVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUQDJA\n0gFJBSQDJB2QVLUMqaNyQXei4CU3bRGbirm90rY2tUVbMHHrQ6YnH3P31q2pYnBRamtPd/ub\nmjalu4P2wiG298wag7S5qWJBd6LKa/Ts9nqiWhxzTdds75k1BmlrS8WC7kTBSzavq7yluNsr\nbV3T2mgLJi5s+8nH3L01TVUMLkrRDkEV229qTncH67xD0GrvmTUGSRR0JwpekudIKp4jqWr5\nOZIISEDyA1LygAQkPyAlD0hA8gNS8oAEJD8gJQ9IQPIDUvKABCQ/ICUPSEDyA1LygAQkPyAl\nD0hA8gNS8oAEJD8gJQ9IQPIDUvKABCQ/ICUPSEDyA1LygAQkPyAlD0hA8gNS8oAEJD8gJQ9I\nQPIDUvKABCQ/ICUPSEDyA1LygAQkPyAlD0hA8ksIKfrvD0gGSDogAQlIQAJSNQEJSH5ASh6Q\ngOQHpOQBCUh+QEoekIDkB6TkAQlIfkBKHpCA5Aek5AEJSH5ASh6QgOQHpOQBCUh+QEoekIDk\nB6TkAQlIfkBKHpCA5Aek5AEJSH5ASh6QgOQHpOQBCUh+QEoekIDkB6TkAQlIfkBKHpCA5Aek\n5AEJSH5ASh6QgOQHpOQBCUh+QEoekIDkB6TkAQlIfkBKHpCA5Aek5AEJSH5ASh6QgOQHpOQB\nCUh+QEoekIDkB6TkAQlIfkBKHpCA5Aek5AEJSH5ASh6QgOQHpOQBCUh+QEoekIDkB6TkAQlI\nfkBKHpCA5Aek5AEJSH5ASh6QgOQHpOQBCUh+QEoekIDkB6TkAQlIfkBKXtRfguvlygOSDEjZ\nDUhA8gNS8oAEJD8gJQ9IQPIDUvKABCQ/ICUPSEDyA1LygAQkPyAlD0hA8gNS8oAEJD8gJQ9I\nQPIDUvKABCQ/ICUPSHdW9AcAABbCSURBVEDyA1LygAQkPyAlD0hA8gNS8oAEJD8gJQ9IQPID\nUvKABCQ/ICUPSEDyA1LygAQkv/qGtH7pvOnTfj3/+S2J1gYSkPzqGNKquy9sKNT/8jkJfg1A\nApJf3ULaOKFfQ9cGze2IuwkgAcmvXiGtPLehtKsjXTOrS0ACkl+dQlp5epmjhoZR7fE2AiQg\n+dUnpA3neHT6jb7vsafnTR1WkHRjvK0ACUh+9QnpZx6ciS2Fn1eO8iYsjrUVIAHJry4hPWfV\nNM7vnLLjBjvp3FgP7oAEJL+6hDTSqpnSddJ27+Hd7DibARKQ/OoR0gprZkj33+zyxvzEC+Ns\nB0hA8qtHSLdbSI+WTB1jp74QYztAApJfPUKybyGdVnpa0FMW0qQY2wESkPzqENKrVsy1pZO3\nnpKffH6MDQEJSH51CGmWhTS3bPpoO31d9A0BCUh+dQhpnAXzatn0yXHfSgISkPzqENJQe753\n+e91oYU0NfqGgAQkvzqEdKp98bt8+nILaVz0DQEJSH71B2mdd4Jq+Yz1dsal0bcEJCD51R8k\n70W7seUzOvrbs4SibwlIQPKrP0jPWki3Bsw5Kz/j1OhbAhKQ/OoP0hMW0uSAOUPsnG2RtwQk\nIPnVH6T5lst9AXMutnOi3yWBBCS/+oP0iOUyPWCOdwJ4S8Cc4IAEJL/6gzTHcnkwYM5wO6cp\n8paABCS/+oM0OxTSZaWQtrU6al2Lqy0F19LUnO4OWtemvf01Ke+geX2621/XtC7dHaz3DvEG\ne8/MAiR3D+2itznuBYpi1nt/kVxVr3+Rope5v0jhLzYMjfliQ/SApAKSKnOQFlsuvwyY8107\nJ41fOJBUQFJlDtKfLJdbAuYMys8YmMY+gaQCkipzkP4a/Lm+nSNtDDmbtfqApAKSKnOQ2iyk\ny8tnrLYzRqaxTyCpgKTKHCTvEdzg8unLLKQJaewSSCogqbIHyb7v2lh+s703mGaksUsgqYCk\nyh6kiRbMc2XTbwyZ7iAgqYCkyh6kBSHvyF6Qn3xSKnd5IKmApMoepLWNgS8qvBX2yVkHAUkF\nJFX2IJlL7NVPSq+7db+FNCuVPQJJBSRVBiHNCDpJaPt5lldzKnsEkgpIqgxCaj05b2ZQ9/ve\n3LD3aV0EJBWQVBmEZCZYND/vOqnJ+y7MF9PZIZBUQFJlEVKTvcx3wz2dU1q8M79/ktIOgaQC\nkiqLkMx0y6bhmre8H3csHmx/Hrg6pf0BSQUkVSYhdXgXzG9oGD5l7sKZE8/2fmhcmNb+gKQC\nkiqTkEzb0Ibygj416yYgqYCkyiYk0zaylNFJ5V/04iwgqYCkyigk0zF9YDdHw19LcWdAUgFJ\nlVVIxmyYMayxoOhb455NdVdAUgFJlV1IO9uwdO4D0x5ctKIj5f0ASQUkVaYh9VRAUgFJBSQD\nJB2QVEAyQNIBSQUkAyQdkFRAMkDSAUkFJAMkHZBUQDJA0gFJBaSdpf0+lelIew+pb58bIPfQ\n5b/rFRKR04BE5CAgETkISEQOAhKRg4BE5CAg7RqtXzpv+rRfz39+S28PZJes5X8W/Obeh+b/\nocJba5fk9jywTiBNCrgohNczvT20ym2+yY5yZugCq+6+sHhT+l8+J+W3IZNU4RZk/6i8fpv/\n2234zthlwQut+vaxDQ11Aml85g9ZcC8OaagIaeOEft1uzaC5qb8TGbNKtyDrR+X1nzR2H9aF\nS8sXKhyCOoF0TcYPWXDbpxSZhEBaeW7Z7bk65fOf4lX5FmT8qMwZUDauxltLz20qHoI6gTS6\n7FeSqUMW3BvD/FEGQ1p5esANGtXew8OskLgF2T4qtweO7L+6L+QfgjqBNCzwl5KVQxZYx6xT\nOkcZCGnDOd7MfqPve+zpeVOLt/HGnh5pWPIWZPqo/KowlMYRd83+7aw7Li0+ypvWdaHiIfjm\n0XUCyX5LzOjeHkWc1v6nd4i+Ew7pZ94SE1sKP68c5U1Y3IPDrJC+BVk+Ki94cBonvF2YsLLw\nQPSUpi5LFQ7B4XvWy8vf9rvTr+ntUcRo0SDvqD08J/Ru+Jx3pOd3Ttlxg510biYe3EW4BRk+\nKjuGeqP/Q5dpv/bUTOicUjgE59XP+0j98zd4fG+PInpTvWM2dKUJvxt6V6ed0nXSdu/B0uwe\nGmWlotyCDB+Vp73hP9VtovcFRKfv8CcUDkH9vCG72d7gSb09jOhd7z1o22bC74Yr7Iwh3V9F\nWm4fkFzYQ6OsVIRbkOWj8mM7tnHdJ7acaqc+V/y5eAjqB1KTvcXT9IJZKX83PMM+4w69G3ov\nKj1aMnWMnfpCDwxRFOEWZPiobLN/LBtWlEz+qZ06o/hj8RDUDyTvn450vto5lXbeDUettf8V\neje071+cVnpa0FNZ+Vc+wi3I8FF5yQ7totLJM+3kO4s/Fg9B/UDynhQu6O1hRO/6/vcXTlEI\nuxu+aqeXfcvuVvuK8/mpD1Cmb0GWj8rLE3586Tknl72RsNCOuPhWkn8I6gfSM/YmP93bw4je\n5JeK/xV2N5xlp5d/+Y33Jue6dIcXIX0Lsn9Uyq7Q4r0EcUPhJ/8Q1A+kBfYm/7m3h5GksLvh\nODv91bLpk+30jLyVZAuFVHtH5bfdHtr5h6B+IM0KfOpYE4XdDYfmJ/cvv6qV9+hjak8MLWKh\nkGrvqPyy28MA/xDUD6Rp9hfQpBfMXmF3Q/tK7JDy5Zfb5ceVz+i1QiHV3lEZ1m3E/iGoH0h3\n2F/App3/terRm68ddenI6+9+emNvDypaIXfDdXbyqPLl19sZl/bE0CIWCqnmjor3j9Swwk+d\nh6B+IP0if5P7mfbH7F9jr5OurYkHFSF3Q+8Vo7Hly3fYd0DO7YmhRSwUUs0dFe/kwYWFnzoP\nQf1AujZ/k09/9ryGbjVe16JX7e1C7obP2sm3BqxwVn7GqT0wsqiFQqq1o/KwHd+FxTOEOg9B\n/UCyrwmXfNrR/sNd/qpX1gq5Gz5hJ08OWMH7ROq2HhhaxEIh1dhRWWmfEzX6rzJ2HoL6gRT6\nwZcBf+ntoalC7obz7eT7Ala42M5J+zL+MQqFVFtHpWmwHdtt/oTOQ1A/kIYUD9F5k5au3bZ1\n1bK7io8nBmf1cUSxkLvhI3by9IAVvLtnhm5WKKSaOirN3lVQRnS+4dB5COoH0hmFxwwL/Snb\n5w70pl3Vi8OKUsjd0Jv8YMAKw+2cDL2oHAqplo7K257x87sI7zwE9QPJO5d3bLfrghSvWpGB\nM6UrFXI3nB0K6bKagVRDR2X52XZU56zuMm12/UHqWJjvDyVTV3gXicnoR52L7boP7WroqDzh\nDer8t7tOrMeHdsHN8P7xW9Xb46hY7Bcbhto5tfBiQ3DZOyodU72XFod1/9epHl9sCG7LmfZX\nkYUPZocXcjdcbCf/MmCF79o5KX+1ZJxiQsrcUdlcuOrJ6JLL2HYegnqH5L21nqnz0soLuRv+\nyU6+JWAFe0WRgT0wsqjFhJS1o7LK+xPfMLb0kjKdh6DuIXlnSmfpvLTyQu6Gf7WTyz7XZ8x2\n+zAk4GzWXisupGwdlVe8lxkayz8R33kI6h7Sy/ZXMbi3h1GxkLthm518efnyq+2MkT0xtIjF\nhZSpo7LEu97JgN+Xz+o8BHUPaZX9VQzq7WFULOxuOCjk3rbMLj+hfEavFRdSlo7KE95L9N95\nJWimfwjqHtKb9rd0Vm8Po2Jhd0P7vmtj+de4eO9uzCib3nvFhZSho/I77ysALgp+V84/BHUP\nybtWTAauFFKhsLvhRDv9ubLpN4ZM773iQsrOUVnkORod8iEp/xDUPaQnMvd8orywu6F3wYPy\nd2QvyE8+KUvf7RIXUmaOyp9PtiP5edhbCf4hqCtILc3l07x/UoJeQ85OYXfDtY2Bd7e37OIB\nn5ztvSpByvJRKXxjy8TQBfxDUC+QXr9h5JBTgu5c9l/vhvnlMzJU6N3wkvz0/qXX3brfLp6p\niy6G3IKsH5XN3mmqN1dYpHgI6gVSi/2Xo1/ZSSdL7G+qMeAfxQwVCsk7labkJKHt9uj3z9RN\nCrkFWT8q3pPNqyt9lWjxENQLpMJ5nL8ombrd/oPScGWvDClyoZBa7SP4Qd1Pqptrlw54n7YX\nC7sF2T4qT9phXFzxyWbxENQNJO/u1VhyovGddmrWL1AY/gzD+5KRn3ed1OQ9rH+xh8YWrbBb\nkOmjssme8TfgjcpLFQ5B3UDa6n2EbMCSLtM6CkdsRG8NKmLhkJq8L5a8p3NKy1A75Sc9NrhI\nhd2CTB8V75p794ulCofg2/UCqfBCZUPjBP8R+TLvg6QNA9/szXFFqMJrXtO9m3DNW96POxZ7\nVxUYuDpo4d5LXLI4k0el7Vv5YZxZ/n53SYVD8Pl96wSSuc67wQ39ht81Z+HjD9303cLPjU/p\ndXuhkxvC6zx7sqP4reDDp8xdOHPi2YWbtLDChnusSLcgu0flngrDb2h4xF/OPwTH1Amk9jGB\nv5GTe/1F1uCiQTJtQwPmB31qtueLdAuye1R+UmH4XSF1HoI6gWTabw24gNrZ2XpW3llESKZt\nZOnck8q/6KVXinYLMntUIkPyD0G9QDLmhctKfh2nT8vshaajQjId0wd2mzn8tV4acWlRb0FG\nj0p0SMVDUD+QjHnl5gv8fwD7j5iZhQMWUmRIxmyYMax4o7417tneGW5A0W9BJo9KDEg7D8Hg\nL32zriDtrHXp47Puu2/mwr9k6JK+Vbdh6dwHpj24aEWlN+EzXa0flUtyexxYZ5CI3Fc3b8gS\npRmQiBwEJCIHAYnIQUAichCQiBwEJCIHAYnIQUAichCQiBwEJCIHAYnIQUAichCQiBwEJCIH\nAameWv2Hxx761fR5Tz23Jv66bc//7uEH7rl/5oJlb7sfWO0HpDrp9bsv+My+Ob+Dj7t42vqo\n666ZPvxfD+xcN7fvPw2++aU0B1uDAakeah5/zG65svb8+m0Rro7QetvX+pavm8t9eGQGrvST\nnYC06/f6sP2CJOQ76NrWyuv+bfi7wtbN7XbiMz1zA2ohIO3qbfnxPqEUdvbuuyqsu/nH76i0\nbm63s4XD+glIu3h//FhFCjs7rfSLyvye+YhaN/cR/ih5AWnXbtLe0kLusJeD1/3Fnnrd3B5T\ne/YGZTUg7dL9tBuYb9/22xXrtm38258fGXPi/l1mHBoo6Ydd1/3Hwbc+tnzt5vbmvy6dNurE\nrjr7zujpG5XJgLQrd1Xn/X3vC7p/vcOmyV/qnPmB18rXHdo5+12XLe0+r+3+r3fO3fO36d2A\n2glIu3B3+3f23c77W/nsuZ/05x/dXjpznD9vr9FBT6L+8DV/gfevTWHstRaQdt2W+I/APrAg\ncIH2S3wMpd/X+kif4pzPPhey+Qn+q4EDHA66VgPSLtvWTxfv6J8P/fK7O4te+nT/GteWQ4vr\nnhL+nu0i/+2pX7sbda0GpF22//T/plQ4F+iW4kIndJv83eLkU3dU2MOivQpLfarSUvURkHbV\n3iw+8vpgU6XFflAk83SXiX/evTDxmK0V93FjceVK7+rWR0CqlXa024rf27L1wUu/ctgBffd6\n96f7jXky6MtcLioc2b6/r7jZTcU3Xft1mXhKYdoBfxWD8l68e/eZj8W4IbtmQKqVrvYO1C/t\nD2su+btcl46Yur108abiw67/ENt92Go7flJz56SXime4TlSDeqFv7qOXPF628zoMSLVS4WHU\n+Px/39X17VTb0atKFh9bmHFgi9rwUX2+OrH7o7/iW7FH6Kc+s16IcRN25YBUK93uHairjem4\nKFfeP5R8QKj4HtEYueGXSj+ot/2Qwrr3uBv9Lh+QaqUp3oEaacywAEe53Mfbui79XGHq3gne\nLF1YWPeQsndpKTQg1Ur3egfqB+Yu7z/2+cSxjV8+Yo9OSZd3Xbr4yC7Je6WjCuuqZ1fUJSDV\nSjO8A3Xucvs26IAZm+zUTY/6Z73t2/VMnuIJPA8k2NNRhXUrv9xH3QJSrTTLO1BnnLjz/z61\nqMuMW4tv+tzcZWLhc627h37WKLxthU9P7McjuxgBqVaa7R2og3b+7wvdefyocAy/3jlpeWHS\n5xPsaGlh3a9VN946C0i10hz/ydAn2rrP2Vh4lW3/zlerpxcWHZJgR5ML6/6wuvHWWUCqlR4u\nOurzVOms4svhf/GnFF9r+EWCHY0prMtpP3ECUq3kQxpcNmtuYc5sf8p/FKbMTbCjIYV1F+lF\nyQ9ItZIPaUnZrKbCnNv9KacWpjybYEffLKwbciUHCgxItVIR0mcD5hXOGLrWn3B8YeHSE4ei\n9MXCulxpK05AqpWKkIYFzPuoN6vzY65FDBsS7OjzhXU5FTVOQKqVipDuDZj3OW/WFf6EIoYk\nbwUd7q3aJ/FI6zIg1UpFSEGnW3/BmzXCn/C5Kv6qFD6h3jdw5tZLQ6r3z8gCqVYqQgp62nNM\nKaTChFxbwMKqigg35EKq99MggFQrFSEFffi7DNJxFdSp/rXS8ysghQSkWqkAac+geWWQTioc\n1mUJdnRiYd1XgmYCKSQg1UoFSHsFzSuDVDzXYU6CHZ1XWPd3QTOBFBKQaqVYkK4pHNbxCXZ0\nVWHdO4NmAikkINVKsSD9qnBYz0mwo+KFjoPesAJSWECqlWJBKn7S/MgEO3q2sO6xkZZuApIN\nSLVSLEjbC5f93i3BN5C3Fy7ktdemKEsDyQtItVIsSMW3aHOTE+ypuO4jURYGkheQaqV4kAI+\nNRu5KwvrnhtlYSB5AalWigdpUeG49nlDb7mj+zdRmMWFdd8V5bEdkLyAVCvFg7T9fYUDe5ne\n8pTc8d0+dLuj+J0uN0UYFpC8gFQrxYPkX0Vyv4rfRZFv89/vXOzErh8XvKKw7oe26WEByQtI\ntVJMSMUXsfUznRHeco2v+lNeLl5E/3o9LCB5AalWignJv0TkboGn+nT2u8Jl8fbr8kp58Wtd\n9ntNDgtIXkCqleJCKl7BO3doxQd3q/6+sFjXq+0vK/5JOqryF40ZIBUDUq0UF5J/FnfuX8K/\nBta0fraw0CHdFvpOcd0zg77DrGv3AskGpFopNqRX/W8d/7fQSze0fqW4zKxu09ccVJx+XsUP\n2XZcV7xeMpCAVBvFhmR+XsSQ+0zIV1iuOLy4xNCSOff7655Y4aucX/mqvxiQgFQbxYdkTvbv\n5u8KvGzqVP/vzpGbS+dd5q/7wbCrTK6+bE9/oW+oh4C7ekCqlRJAaj3Cv6Pnjp5futKSzj8n\nHyo//2H7iZ3rfuN/Ava5ZMjenUucUe9/kIBUMyWAZFZ9svO+nvv02Oc757zx30d1zjlkecA2\nN3+9y7pHT+z24LB17oh/7DJ39yvr/e8RkGqnJJDMW/+c69rBx1/04xvGX3XxaYd1nfqh5wM2\nacyWAd3W/dBJw8ffMe3uieMu6/fpPt3mHPSo21takwGpVkoEyWw6Paf69+awXV7XR668s7NX\nu7uRtRuQaqVkkIyZdEBFBnuNqXBpxyf/STL658fd3LxaD0i1UlJI5u3vVPjDMuDVwHWKtd/w\n3oqMjv5N9Tds1whItVJiSMa8PHjvQAZ79HtC7nbT+I+EKXrP956s5gbtWgGpVqoCkjHr//u4\nfUodHPmziNdhffIH5ZZ2O/LiWRE+ZFE/Aale2vr7cRd87aPve2efvvse8tl+V9z7VpyV355+\n5Vn/+qGD9t19r/cc9umvff+/Hl6T1jBrNSAROQhIRA4CEpGDgETkICAROQhIRA4CEpGDgETk\nICAROQhIRA4CEpGDgETkICAROQhIRA4CEpGDgETkICAROQhIRA4CEpGDgETkICAROagI6agR\nRJS4Lt9JQERERES13/8BfyLBauXcQtoAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "dmrs %>%\n",
    "    ggplot(aes(x = nCG)) +\n",
    "    geom_histogram() +\n",
    "    wiscR::light_theme() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbd2240a",
   "metadata": {},
   "source": [
    "# callDMR\n",
    "\n",
    "Parameters that we can vary: \n",
    "- `p.threshold`\n",
    "- `minlen` is the minimum length in bp to be considered a region (default 50)\n",
    "- `minCG` (default 3)\n",
    "- `dis.merge` if two DMRs are dis.merge or fewer basepairs apart then they are merged.\n",
    "- `pct.sig` percentage of CG called significant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "005f0cbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "load.samples <- filt.df %>% filter(cohort == \"AD\") %>% pull(sample) %>% as.character()\n",
    "control.samples <- filt.df %>% filter(cohort == \"CONTROL\") %>% pull(sample) %>% as.character()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "e831092d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "33"
      ],
      "text/latex": [
       "33"
      ],
      "text/markdown": [
       "33"
      ],
      "text/plain": [
       "[1] 33"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "28"
      ],
      "text/latex": [
       "28"
      ],
      "text/markdown": [
       "28"
      ],
      "text/plain": [
       "[1] 28"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(load.samples)\n",
    "length(control.samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "82da77fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "prepare_granges <- function(DT, chr=CHR){\n",
    "    # Munges dataframe (changes column names)\n",
    "    # to be amenable to GRanges() constructor\n",
    "    DT.2 <- DT\n",
    "    \n",
    "    colnames(DT.2)[colnames(DT.2) == \"pos\"] <- \"start\"\n",
    "    DT.2$end <- DT.2$start + 2\n",
    "    DT.2$chr <- chr\n",
    "    return(DT.2)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "make_difference_track <- function(Meth, load.samples, control.samples, chr = CHR){\n",
    "    \n",
    "    # Compute mean difference\n",
    "    meandiff <- rowMeans(Meth[ , load.samples, with = FALSE], na.rm = TRUE) - \n",
    "            rowMeans(Meth[ , control.samples, with = FALSE], na.rm = TRUE)\n",
    "    # Construct GRanges\n",
    "    dt <- data.table(diff=meandiff, chr=CHR, start=Meth$start, end=Meth$end) %>%\n",
    "            drop_na() # important to prevent errors\n",
    "    gr <- GRanges(dt)\n",
    "    track <- DataTrack(gr, type=c(\"p\", \"smooth\"), baseline=0, col=colors[2],\n",
    "                       col.baseline=\"black\", name=\"Mean Methylation Difference\", \n",
    "                       fontsize = FS)\n",
    "    return(track)\n",
    "} \n",
    "\n",
    "\n",
    "make_methylation_track <- function(Meth, group){\n",
    "    # Meth matrix should have been run through \n",
    "    DataTrack(GRanges(Meth), \n",
    "              groups = group, \n",
    "              name = \"Methylation\",                             \n",
    "              type = c(\"a\", \"p\"), \n",
    "              legend = TRUE) %>%\n",
    "    return()\n",
    "}\n",
    "\n",
    "\n",
    "make_cpg_track <- function(start, stop){\n",
    "    cpgIslands <- UcscTrack(genome = GEN, chromosome = CHR, \n",
    "                        track = \"cpgIslandExt\", \n",
    "                        from = start, to = stop,\n",
    "                        trackType = \"AnnotationTrack\", \n",
    "                        start = \"chromStart\", end = \"chromEnd\", \n",
    "                        id = \"name\", shape = \"box\", fill = colors[4], \n",
    "                        name = \"CpG Islands\", fontsize = FS)\n",
    "    return(cpgIslands)\n",
    "}\n",
    "\n",
    "\n",
    "make_snp_track <- function(start, stop){\n",
    "    \n",
    "    snpLocations <-  UcscTrack(genome = GEN, chromosome = CHR, \n",
    "                           track = \"snp141\", \n",
    "                           from = start, to = stop,\n",
    "                           trackType = \"AnnotationTrack\", \n",
    "                           start = \"chromStart\", end = \"chromEnd\", \n",
    "                           id = \"name\", feature = \"func\", \n",
    "                           strand = \"strand\", shape = \"box\", \n",
    "                           stacking = \"dense\", fill = colors[2],\n",
    "                           name = \"SNPs\", fontsize = FS)\n",
    "    return(snpLocations)\n",
    "}\n",
    "\n",
    "make_p_track <- function(test.cohort){\n",
    "     # Meth matrix should have been run through \n",
    "\n",
    "    data.frame(chr = test.cohort$chr, start = test.cohort$pos, stop = test.cohort$pos + 2, \n",
    "               log.p = -1*log10(test.cohort$pvals)) %>%\n",
    "        drop_na() %>%\n",
    "        GRanges() %>%\n",
    "        DataTrack(name = \"-log10(p)\", type = c(\"smooth\", \"p\"), \n",
    "              legend = TRUE, col=colors[1], fontsize = FS) %>%\n",
    "    return()\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "492b214c",
   "metadata": {},
   "outputs": [],
   "source": [
    "Meth.2 <- prepare_granges(Meth)\n",
    "group <- filt.df$cohort[match(colnames(Meth)[-1], filt.df$sample)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4eb05c66",
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"Simple\" tracks without too many parameters\n",
    "dmrs.track <- AnnotationTrack(GRanges(dmrs), genome = GEN, name = \"DMRs\", fontsize = FS,\n",
    "                             fill = colors[3], col = \"white\")\n",
    "ideogram.track <- IdeogramTrack(genome = GEN, chromosome = CHR, fontsize = FS - 2)\n",
    "axis.track <- GenomeAxisTrack()\n",
    "p.track <- make_p_track(test.cohort)\n",
    "\n",
    "# Tracks from custom functions\n",
    "diff.track <- make_difference_track(Meth.2, load.samples, control.samples)\n",
    "\n",
    "# Show up on top of genomic information\n",
    "top.list <- list(ideogram.track, axis.track)\n",
    "\n",
    "# Show up below gneomic info\n",
    "bottom.list <- list(dmrs.track, p.track, diff.track)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "02468ac9",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_all_tracks <- function(start, stop, top.tracks.list, bottom.tracks.list, \n",
    "                            odir, pad=c(100,1000, 10000)){\n",
    "    # start : starting position\n",
    "    # stop: stopping position\n",
    "    # tracks_list: list of tracks to plot, produced beforehand\n",
    "    # odir: where to save pngs\n",
    "    # pad: vector specifying how big a window\n",
    "    \n",
    "    # File setup\n",
    "    dir.create(odir, showWarnings=FALSE)\n",
    "    \n",
    "    # Make the genes track based on the largest window\n",
    "    a <- start - max(pad)\n",
    "    b <- stop + max(pad)\n",
    "    \n",
    "    # All the tracks\n",
    "    ref.genes.track <- make_gene_track(start = a, stop = b)\n",
    "    #snp.track <- make_snp_track(start = a, stop = b)\n",
    "    cpg.track <- make_cpg_track(start = a, stop = b)\n",
    "    \n",
    "    # Cobble together\n",
    "    mid.tracks.list <- list(ref.genes.track, cpg.track)\n",
    "        \n",
    "    # Loop thru padding values\n",
    "    for (p in pad){\n",
    "        a <- start - p\n",
    "        b <- stop  + p\n",
    "        \n",
    "        tracks.list <- append(top.tracks.list, mid.tracks.list)\n",
    "        tracks.list <- append(tracks.list, bottom.tracks.list)\n",
    "        \n",
    "        \n",
    "        # Save, how do we feel about default sizes?\n",
    "        pdf(paste0(odir, as.character(start), \"-\", as.character(p), \"-DMRs.pdf\"))\n",
    "        plotTracks(tracks.list, from = a, to = b, \n",
    "                   background.title = colors[6])\n",
    "        dev.off()\n",
    "    }    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9caaa364",
   "metadata": {},
   "outputs": [],
   "source": [
    "ix <- 1\n",
    "a <- dmrs$start[ix]\n",
    "b <- dmrs$end[ix]\n",
    "\n",
    "tmp <- make_gene_track(a, b)\n",
    "\n",
    "#plot_all_tracks(a, b, top.list, bottom.list, file.path(idir, \"figs/\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c88c2954",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_by_ix <- function(ix){\n",
    "    plot_all_tracks(dmrs$start[ix], dmrs$end[ix], top.list, bottom.list, file.path(idir, \"figs/\"))\n",
    "}\n",
    "\n",
    "mclapply(1:nrow(dmrs), plot_by_ix, mc.cores = 8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11aa2f3c",
   "metadata": {},
   "source": [
    "# Regions of Interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "9ce3c810",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in assign(i, as.numeric(val)):\n",
      "\"NAs introduced by coercion\"\n",
      "Warning message in assign(i, as.numeric(val)):\n",
      "\"NAs introduced by coercion\"\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in .fillWithDefaults(data.frame(start = as.integer(start), end = as.integer(end)), : Number of elements in argument 'transcript' is invalid\n",
     "output_type": "error",
     "traceback": [
      "Error in .fillWithDefaults(data.frame(start = as.integer(start), end = as.integer(end)), : Number of elements in argument 'transcript' is invalid\nTraceback:\n",
      "1. make_gene_track(a, b)",
      "2. UcscTrack(track = \"knownGene\", type = \"GeneRegionTrack\", genome = gen, \n .     chromosome = chr, from = start, to = stop, trackType = \"GeneRegionTrack\", \n .     rstarts = \"exonStarts\", rends = \"exonEnds\", gene = \"name\", \n .     symbol = \"name\", transcript = \"name\", strand = \"strand\", \n .     fill = \"#8282d2\", name = \"UCSC Genes\")   # at line 22-29 of file <text>",
      "3. do.call(trackType, args = c(list(chromosome = chromosome, genome = genome, \n .     name = name), args))",
      "4. do.call(trackType, args = c(list(chromosome = chromosome, genome = genome, \n .     name = name), args))",
      "5. GeneRegionTrack(chromosome = \"chr18\", genome = \"hg38\", name = \"UCSC Genes\", \n .     type = \"GeneRegionTrack\", rstarts = \"exonStarts\", rends = \"exonEnds\", \n .     gene = c(\"ENST00000322342.4\", \"ENST00000537114.2\", \"ENST00000582437.1\", \n .     \"ENST00000581620.1\", \"ENST00000322038.5\", \"ENST00000580243.3\", \n .     \"ENST00000560918.2\"), symbol = c(\"ENST00000322342.4\", \"ENST00000537114.2\", \n .     \"ENST00000582437.1\", \"ENST00000581620.1\", \"ENST00000322038.5\", \n .     \"ENST00000580243.3\", \"ENST00000560918.2\"), transcript = c(\"ENST00000322342.4\", \n .     \"ENST00000537114.2\", \"ENST00000582437.1\", \"ENST00000581620.1\", \n .     \"ENST00000322038.5\", \"ENST00000580243.3\", \"ENST00000560918.2\"\n .     ), strand = c(\"-\", \"-\", \"-\", \"-\", \"+\", \"+\", \"+\"), fill = \"#8282d2\", \n .     start = 75165108, end = 75239506)",
      "6. .buildRange(range = range, groupId = \"transcript\", start = rstarts, \n .     end = rends, width = rwidths, args = args, defaults = defs, \n .     chromosome = chromosome, tstart = start, tend = end, trackType = \"GeneRegionTrack\", \n .     importFun = importFunction, genome = genome)",
      "7. .buildRange(range = range, groupId = \"transcript\", start = rstarts, \n .     end = rends, width = rwidths, args = args, defaults = defs, \n .     chromosome = chromosome, tstart = start, tend = end, trackType = \"GeneRegionTrack\", \n .     importFun = importFunction, genome = genome)",
      "8. .local(range, start, end, width, ...)",
      "9. .buildRange(start = start, end = end, width = width, asIRanges = asIRanges, \n .     by = by, len = len, ...)",
      "10. .buildRange(start = start, end = end, width = width, asIRanges = asIRanges, \n  .     by = by, len = len, ...)",
      "11. .local(range, start, end, width, ...)",
      "12. .fillWithDefaults(data.frame(start = as.integer(start), end = as.integer(end)), \n  .     defaults, args, len, by)",
      "13. stop(\"Number of elements in argument '\", a, \"' is invalid\")"
     ]
    }
   ],
   "source": [
    "roi.df <-roi.df %>% filter(chr == CHR)\n",
    "a <- roi.df$start[1] - 30000\n",
    "b <- roi.df$end[1] + 30000\n",
    "\n",
    "\n",
    "make_gene_track <- function(start, stop, gen=GEN, chr=CHR){\n",
    "    # Produces the track with exons, \n",
    "    \n",
    "    out <- UcscTrack(track = \"GENCODE V38\", name = \"something\",type=\"GeneRegionTrack\",\n",
    "                      genome = gen, chromosome = chr,\n",
    "                      from = start, to = stop,\n",
    "                      rstarts=\"thickStart\", rends=\"thickEnd\", \n",
    "                      gene=\"name\", symbol=\"name\", \n",
    "                      transcript=\"name\", strand=\"strand\",\n",
    "                      transcriptAnnotation = \"name\", showID=TRUE)\n",
    "    return(out)\n",
    "} \n",
    "\n",
    "make_gene_track <- function(start, stop, gen=GEN, chr=CHR){\n",
    "    # Produces the track with exons, \n",
    "    \n",
    "    out <- UcscTrack(track = \"knownGene\", type=\"GeneRegionTrack\",\n",
    "                      genome = gen, chromosome = chr,\n",
    "                      from = start, to = stop,\n",
    "                      trackType = \"GeneRegionTrack\", \n",
    "                        rstarts = \"exonStarts\", rends = \"exonEnds\", \n",
    "                        gene = \"name\", symbol = \"name\", \n",
    "                        transcript = \"name\", strand = \"strand\", \n",
    "                        fill = \"#8282d2\", name = \"UCSC Genes\")\n",
    "                    #transcriptAnnotation = \"name2\")\n",
    "    return(out)\n",
    "} \n",
    "\n",
    "\n",
    "\n",
    "tmp <- make_gene_track(a,b)\n",
    "plotTracks(tmp,a,b)\n",
    "\n",
    "#plot_all_tracks(, top.list, bottom.list, \n",
    " #               file.path(idir, \"figs/\", paste0(roi.df$name[1], \"/\")),\n",
    "  #              pad = c(1000, 3000, 5000))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b239c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "f8263e68",
   "metadata": {},
   "outputs": [],
   "source": [
    "from <- 65921878\n",
    "to <- 65980988\n",
    "\n",
    "knownGenes <- UcscTrack(genome=\"mm9\", chromosome=\"chrX\", track=\"knownGene\",\n",
    "from=from, to=to, trackType=\"GeneRegionTrack\",\n",
    "rstarts=\"exonStarts\", rends=\"exonEnds\", gene=\"name\",\n",
    "symbol=\"name\", transcript=\"name\", strand=\"strand\",\n",
    "fill=\"#8282d2\", name=\"UCSC Genes\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "c2593612",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'GeneRegionTrack' [package \"Gviz\"] with 10 slots\n",
      "  ..@ start     : num 65921878\n",
      "  ..@ end       : num 6.6e+07\n",
      "  ..@ stacking  : chr \"squish\"\n",
      "  ..@ stacks    : num(0) \n",
      "  ..@ range     :Formal class 'GRanges' [package \"GenomicRanges\"] with 7 slots\n",
      "  .. .. ..@ seqnames       :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. ..@ values         : Factor w/ 1 level \"chrX\": 1\n",
      "  .. .. .. .. ..@ lengths        : int 76\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ ranges         :Formal class 'IRanges' [package \"IRanges\"] with 6 slots\n",
      "  .. .. .. .. ..@ start          : int [1:76] 65920688 65926493 65930188 65931354 65931729 65941456 65944675 65931729 65941456 65944675 ...\n",
      "  .. .. .. .. ..@ width          : int [1:76] 2355 58 113 221 314 54 995 314 54 95 ...\n",
      "  .. .. .. .. ..@ NAMES          : NULL\n",
      "  .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ strand         :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. ..@ values         : Factor w/ 3 levels \"+\",\"-\",\"*\": 2 1\n",
      "  .. .. .. .. ..@ lengths        : int [1:2] 4 72\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ seqinfo        :Formal class 'Seqinfo' [package \"GenomeInfoDb\"] with 4 slots\n",
      "  .. .. .. .. ..@ seqnames   : chr \"chrX\"\n",
      "  .. .. .. .. ..@ seqlengths : int NA\n",
      "  .. .. .. .. ..@ is_circular: logi NA\n",
      "  .. .. .. .. ..@ genome     : chr \"mm9\"\n",
      "  .. .. ..@ elementMetadata:Formal class 'DFrame' [package \"S4Vectors\"] with 6 slots\n",
      "  .. .. .. .. ..@ rownames       : NULL\n",
      "  .. .. .. .. ..@ nrows          : int 76\n",
      "  .. .. .. .. ..@ listData       :List of 7\n",
      "  .. .. .. .. .. ..$ feature   : chr [1:76] \"exon_1\" \"exon_1\" \"exon_1\" \"exon_1\" ...\n",
      "  .. .. .. .. .. ..$ id        : chr [1:76] \"exon_1\" \"exon_1\" \"exon_1\" \"exon_1\" ...\n",
      "  .. .. .. .. .. ..$ exon      : chr [1:76] \"exon_1\" \"exon_1\" \"exon_1\" \"exon_1\" ...\n",
      "  .. .. .. .. .. ..$ transcript: chr [1:76] \"uc012hjf.1\" \"uc012hjf.1\" \"uc012hjf.1\" \"uc012hjf.1\" ...\n",
      "  .. .. .. .. .. ..$ gene      : chr [1:76] \"uc012hjf.1\" \"uc012hjf.1\" \"uc012hjf.1\" \"uc012hjf.1\" ...\n",
      "  .. .. .. .. .. ..$ symbol    : chr [1:76] \"uc012hjf.1\" \"uc012hjf.1\" \"uc012hjf.1\" \"uc012hjf.1\" ...\n",
      "  .. .. .. .. .. ..$ density   : num [1:76] 1 1 1 1 1 1 1 1 1 1 ...\n",
      "  .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ chromosome: chr \"chrX\"\n",
      "  ..@ genome    : Named chr \"mm9\"\n",
      "  .. ..- attr(*, \"names\")= chr \"chrX\"\n",
      "  ..@ dp        :Formal class 'DisplayPars' [package \"Gviz\"] with 1 slot\n",
      "  .. .. ..@ pars:List of 72\n",
      "  .. .. .. ..$ arrowHeadWidth      : num 10\n",
      "  .. .. .. ..$ arrowHeadMaxWidth   : num 20\n",
      "  .. .. .. ..$ col                 : chr \"darkgray\"\n",
      "  .. .. .. ..$ collapseTranscripts : logi FALSE\n",
      "  .. .. .. ..$ exonAnnotation      : NULL\n",
      "  .. .. .. ..$ fill                : chr \"#8282d2\"\n",
      "  .. .. .. ..$ min.distance        : num 0\n",
      "  .. .. .. ..$ shape               : chr [1:2] \"smallArrow\" \"box\"\n",
      "  .. .. .. ..$ showExonId          : NULL\n",
      "  .. .. .. ..$ thinBoxFeature      : chr [1:10] \"utr\" \"ncRNA\" \"utr3\" \"utr5\" ...\n",
      "  .. .. .. ..$ transcriptAnnotation: NULL\n",
      "  .. .. .. ..$ .__appliedScheme    : chr \"default\"\n",
      "  .. .. .. ..$ cex.group           : num 0.6\n",
      "  .. .. .. ..$ cex                 : num 1\n",
      "  .. .. .. ..$ col.line            : chr \"darkgray\"\n",
      "  .. .. .. ..$ featureAnnotation   : NULL\n",
      "  .. .. .. ..$ fontfamily.group    : chr \"sans\"\n",
      "  .. .. .. ..$ fontcolor.group     : chr \"#808080\"\n",
      "  .. .. .. ..$ fontcolor.item      : chr \"white\"\n",
      "  .. .. .. ..$ fontface.group      : num 2\n",
      "  .. .. .. ..$ fontsize.group      : num 12\n",
      "  .. .. .. ..$ groupAnnotation     : NULL\n",
      "  .. .. .. ..$ just.group          : chr \"left\"\n",
      "  .. .. .. ..$ lex                 : num 1\n",
      "  .. .. .. ..$ lineheight          : num 1\n",
      "  .. .. .. ..$ lty                 : chr \"solid\"\n",
      "  .. .. .. ..$ lwd                 : num 1\n",
      "  .. .. .. ..$ mergeGroups         : logi FALSE\n",
      "  .. .. .. ..$ min.height          : num 3\n",
      "  .. .. .. ..$ min.width           : num 1\n",
      "  .. .. .. ..$ rotation            : num 0\n",
      "  .. .. .. ..$ rotation.group      : num 0\n",
      "  .. .. .. ..$ rotation.item       : num 0\n",
      "  .. .. .. ..$ showFeatureId       : NULL\n",
      "  .. .. .. ..$ showId              : NULL\n",
      "  .. .. .. ..$ showOverplotting    : logi FALSE\n",
      "  .. .. .. ..$ size                : num 1\n",
      "  .. .. .. ..$ stackHeight         : num 0.75\n",
      "  .. .. .. ..$ reverseStacking     : logi FALSE\n",
      "  .. .. .. ..$ alpha               : num 1\n",
      "  .. .. .. ..$ alpha.title         : NULL\n",
      "  .. .. .. ..$ background.panel    : chr \"transparent\"\n",
      "  .. .. .. ..$ background.title    : chr \"lightgray\"\n",
      "  .. .. .. ..$ background.legend   : chr \"transparent\"\n",
      "  .. .. .. ..$ cex.axis            : NULL\n",
      "  .. .. .. ..$ cex.title           : NULL\n",
      "  .. .. .. ..$ col.axis            : chr \"white\"\n",
      "  .. .. .. ..$ col.border.title    : chr \"white\"\n",
      "  .. .. .. ..$ col.frame           : chr \"lightgray\"\n",
      "  .. .. .. ..$ col.grid            : chr \"#808080\"\n",
      "  .. .. .. ..$ col.symbol          : NULL\n",
      "  .. .. .. ..$ fontcolor.title     : chr \"white\"\n",
      "  .. .. .. ..$ collapse            : logi TRUE\n",
      "  .. .. .. ..$ fontcolor           : chr \"black\"\n",
      "  .. .. .. ..$ fontface.title      : num 2\n",
      "  .. .. .. ..$ fontface            : num 1\n",
      "  .. .. .. ..$ fontfamily.title    : chr \"sans\"\n",
      "  .. .. .. ..$ fontfamily          : chr \"sans\"\n",
      "  .. .. .. ..$ fontsize            : num 12\n",
      "  .. .. .. ..$ frame               : logi FALSE\n",
      "  .. .. .. ..$ grid                : logi FALSE\n",
      "  .. .. .. ..$ h                   : num -1\n",
      "  .. .. .. ..$ lty.grid            : chr \"solid\"\n",
      "  .. .. .. ..$ lwd.border.title    : num 1\n",
      "  .. .. .. ..$ lwd.title           : num 1\n",
      "  .. .. .. ..$ lwd.grid            : num 1\n",
      "  .. .. .. ..$ reverseStrand       : logi FALSE\n",
      "  .. .. .. ..$ rotation.title      : num 90\n",
      "  .. .. .. ..$ showAxis            : logi TRUE\n",
      "  .. .. .. ..$ showTitle           : logi TRUE\n",
      "  .. .. .. ..$ v                   : num -1\n",
      "  .. .. .. ..$ lwd.border          : num 1\n",
      "  ..@ name      : chr \"UCSC Genes\"\n",
      "  ..@ imageMap  : NULL\n"
     ]
    }
   ],
   "source": [
    "str(knownGenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ac93766c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "76"
      ],
      "text/latex": [
       "76"
      ],
      "text/markdown": [
       "76"
      ],
      "text/plain": [
       "[1] 76"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(identifier(knownGenes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "7ff74b70",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formal class 'GeneRegionTrack' [package \"Gviz\"] with 10 slots\n",
      "  ..@ start     : num 75195108\n",
      "  ..@ end       : num 75209506\n",
      "  ..@ stacking  : chr \"squish\"\n",
      "  ..@ stacks    : num(0) \n",
      "  ..@ range     :Formal class 'GRanges' [package \"GenomicRanges\"] with 7 slots\n",
      "  .. .. ..@ seqnames       :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. ..@ values         : Factor w/ 1 level \"chr18\": 1\n",
      "  .. .. .. .. ..@ lengths        : int 4\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ ranges         :Formal class 'IRanges' [package \"IRanges\"] with 6 slots\n",
      "  .. .. .. .. ..@ start          : int [1:4] 75201415 75201415 75201716 75201877\n",
      "  .. .. .. .. ..@ width          : int [1:4] 7644 766 465 304\n",
      "  .. .. .. .. ..@ NAMES          : NULL\n",
      "  .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ strand         :Formal class 'Rle' [package \"S4Vectors\"] with 4 slots\n",
      "  .. .. .. .. ..@ values         : Factor w/ 3 levels \"+\",\"-\",\"*\": 2\n",
      "  .. .. .. .. ..@ lengths        : int 4\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ seqinfo        :Formal class 'Seqinfo' [package \"GenomeInfoDb\"] with 4 slots\n",
      "  .. .. .. .. ..@ seqnames   : chr \"chr18\"\n",
      "  .. .. .. .. ..@ seqlengths : int NA\n",
      "  .. .. .. .. ..@ is_circular: logi NA\n",
      "  .. .. .. .. ..@ genome     : chr \"hg38\"\n",
      "  .. .. ..@ elementMetadata:Formal class 'DFrame' [package \"S4Vectors\"] with 6 slots\n",
      "  .. .. .. .. ..@ rownames       : NULL\n",
      "  .. .. .. .. ..@ nrows          : int 4\n",
      "  .. .. .. .. ..@ listData       :List of 7\n",
      "  .. .. .. .. .. ..$ feature   : chr [1:4] \"exon_1\" \"exon_1\" \"exon_1\" \"exon_1\"\n",
      "  .. .. .. .. .. ..$ id        : chr [1:4] \"exon_1\" \"exon_1\" \"exon_1\" \"exon_1\"\n",
      "  .. .. .. .. .. ..$ exon      : chr [1:4] \"exon_1\" \"exon_1\" \"exon_1\" \"exon_1\"\n",
      "  .. .. .. .. .. ..$ transcript: chr [1:4] \"ENST00000322342.4\" \"ENST00000537114.2\" \"ENST00000582437.1\" \"ENST00000581620.1\"\n",
      "  .. .. .. .. .. ..$ gene      : chr [1:4] \"ENST00000322342.4\" \"ENST00000537114.2\" \"ENST00000582437.1\" \"ENST00000581620.1\"\n",
      "  .. .. .. .. .. ..$ symbol    : chr [1:4] \"ENST00000322342.4\" \"ENST00000537114.2\" \"ENST00000582437.1\" \"ENST00000581620.1\"\n",
      "  .. .. .. .. .. ..$ density   : num [1:4] 1 1 1 1\n",
      "  .. .. .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. .. .. ..@ elementMetadata: NULL\n",
      "  .. .. .. .. ..@ metadata       : list()\n",
      "  .. .. ..@ elementType    : chr \"ANY\"\n",
      "  .. .. ..@ metadata       : list()\n",
      "  ..@ chromosome: chr \"chr18\"\n",
      "  ..@ genome    : Named chr \"hg38\"\n",
      "  .. ..- attr(*, \"names\")= chr \"chr18\"\n",
      "  ..@ dp        :Formal class 'DisplayPars' [package \"Gviz\"] with 1 slot\n",
      "  .. .. ..@ pars:List of 74\n",
      "  .. .. .. ..$ arrowHeadWidth      : num 10\n",
      "  .. .. .. ..$ arrowHeadMaxWidth   : num 20\n",
      "  .. .. .. ..$ col                 : chr \"darkgray\"\n",
      "  .. .. .. ..$ collapseTranscripts : logi FALSE\n",
      "  .. .. .. ..$ exonAnnotation      : NULL\n",
      "  .. .. .. ..$ fill                : chr \"#FFD58A\"\n",
      "  .. .. .. ..$ min.distance        : num 0\n",
      "  .. .. .. ..$ shape               : chr [1:2] \"smallArrow\" \"box\"\n",
      "  .. .. .. ..$ showExonId          : NULL\n",
      "  .. .. .. ..$ thinBoxFeature      : chr [1:10] \"utr\" \"ncRNA\" \"utr3\" \"utr5\" ...\n",
      "  .. .. .. ..$ transcriptAnnotation: chr [1:4] \"ZADH2\" \"ZADH2\" \"ZADH2\" \"ZADH2\"\n",
      "  .. .. .. ..$ .__appliedScheme    : chr \"default\"\n",
      "  .. .. .. ..$ cex.group           : num 0.6\n",
      "  .. .. .. ..$ cex                 : num 1\n",
      "  .. .. .. ..$ col.line            : chr \"darkgray\"\n",
      "  .. .. .. ..$ featureAnnotation   : NULL\n",
      "  .. .. .. ..$ fontfamily.group    : chr \"sans\"\n",
      "  .. .. .. ..$ fontcolor.group     : chr \"#808080\"\n",
      "  .. .. .. ..$ fontcolor.item      : chr \"white\"\n",
      "  .. .. .. ..$ fontface.group      : num 2\n",
      "  .. .. .. ..$ fontsize.group      : num 12\n",
      "  .. .. .. ..$ groupAnnotation     : NULL\n",
      "  .. .. .. ..$ just.group          : chr \"left\"\n",
      "  .. .. .. ..$ lex                 : num 1\n",
      "  .. .. .. ..$ lineheight          : num 1\n",
      "  .. .. .. ..$ lty                 : chr \"solid\"\n",
      "  .. .. .. ..$ lwd                 : num 1\n",
      "  .. .. .. ..$ mergeGroups         : logi FALSE\n",
      "  .. .. .. ..$ min.height          : num 3\n",
      "  .. .. .. ..$ min.width           : num 1\n",
      "  .. .. .. ..$ rotation            : num 0\n",
      "  .. .. .. ..$ rotation.group      : num 0\n",
      "  .. .. .. ..$ rotation.item       : num 0\n",
      "  .. .. .. ..$ showFeatureId       : NULL\n",
      "  .. .. .. ..$ showId              : NULL\n",
      "  .. .. .. ..$ showOverplotting    : logi FALSE\n",
      "  .. .. .. ..$ size                : num 1\n",
      "  .. .. .. ..$ stackHeight         : num 0.75\n",
      "  .. .. .. ..$ reverseStacking     : logi FALSE\n",
      "  .. .. .. ..$ alpha               : num 1\n",
      "  .. .. .. ..$ alpha.title         : NULL\n",
      "  .. .. .. ..$ background.panel    : chr \"transparent\"\n",
      "  .. .. .. ..$ background.title    : chr \"lightgray\"\n",
      "  .. .. .. ..$ background.legend   : chr \"transparent\"\n",
      "  .. .. .. ..$ cex.axis            : NULL\n",
      "  .. .. .. ..$ cex.title           : NULL\n",
      "  .. .. .. ..$ col.axis            : chr \"white\"\n",
      "  .. .. .. ..$ col.border.title    : chr \"white\"\n",
      "  .. .. .. ..$ col.frame           : chr \"lightgray\"\n",
      "  .. .. .. ..$ col.grid            : chr \"#808080\"\n",
      "  .. .. .. ..$ col.symbol          : NULL\n",
      "  .. .. .. ..$ fontcolor.title     : chr \"white\"\n",
      "  .. .. .. ..$ collapse            : logi TRUE\n",
      "  .. .. .. ..$ fontcolor           : chr \"black\"\n",
      "  .. .. .. ..$ fontface.title      : num 2\n",
      "  .. .. .. ..$ fontface            : num 1\n",
      "  .. .. .. ..$ fontfamily.title    : chr \"sans\"\n",
      "  .. .. .. ..$ fontfamily          : chr \"sans\"\n",
      "  .. .. .. ..$ fontsize            : num 18\n",
      "  .. .. .. ..$ frame               : logi FALSE\n",
      "  .. .. .. ..$ grid                : logi FALSE\n",
      "  .. .. .. ..$ h                   : num -1\n",
      "  .. .. .. ..$ lty.grid            : chr \"solid\"\n",
      "  .. .. .. ..$ lwd.border.title    : num 1\n",
      "  .. .. .. ..$ lwd.title           : num 1\n",
      "  .. .. .. ..$ lwd.grid            : num 1\n",
      "  .. .. .. ..$ reverseStrand       : logi FALSE\n",
      "  .. .. .. ..$ rotation.title      : num 90\n",
      "  .. .. .. ..$ showAxis            : logi TRUE\n",
      "  .. .. .. ..$ showTitle           : logi TRUE\n",
      "  .. .. .. ..$ v                   : num -1\n",
      "  .. .. .. ..$ lwd.border          : num 1\n",
      "  .. .. .. ..$ color               : chr \"#E18727FF\"\n",
      "  .. .. .. ..$ showID              : logi FALSE\n",
      "  ..@ name      : chr \"UCSC\"\n",
      "  ..@ imageMap  : NULL\n"
     ]
    }
   ],
   "source": [
    "str(tmp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "3f56a12e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<pre class=language-r><code>new(\"standardGeneric\", .Data = function (GdObject, ...) \n",
       "standardGeneric(\"identifier\"), generic = structure(\"identifier\", package = \"Gviz\"), \n",
       "<span style=white-space:pre-wrap>    package = \"Gviz\", group = list(), valueClass = character(0), </span>\n",
       "<span style=white-space:pre-wrap>    signature = \"GdObject\", default = NULL, skeleton = (function (GdObject, </span>\n",
       "<span style=white-space:pre-wrap>        ...) </span>\n",
       "<span style=white-space:pre-wrap>    stop(\"invalid call in method dispatch to 'identifier' (no default method)\", </span>\n",
       "<span style=white-space:pre-wrap>        domain = NA))(GdObject, ...))</span></code></pre>"
      ],
      "text/latex": [
       "\\begin{minted}{r}\n",
       "new(\"standardGeneric\", .Data = function (GdObject, ...) \n",
       "standardGeneric(\"identifier\"), generic = structure(\"identifier\", package = \"Gviz\"), \n",
       "    package = \"Gviz\", group = list(), valueClass = character(0), \n",
       "    signature = \"GdObject\", default = NULL, skeleton = (function (GdObject, \n",
       "        ...) \n",
       "    stop(\"invalid call in method dispatch to 'identifier' (no default method)\", \n",
       "        domain = NA))(GdObject, ...))\n",
       "\\end{minted}"
      ],
      "text/markdown": [
       "```r\n",
       "new(\"standardGeneric\", .Data = function (GdObject, ...) \n",
       "standardGeneric(\"identifier\"), generic = structure(\"identifier\", package = \"Gviz\"), \n",
       "    package = \"Gviz\", group = list(), valueClass = character(0), \n",
       "    signature = \"GdObject\", default = NULL, skeleton = (function (GdObject, \n",
       "        ...) \n",
       "    stop(\"invalid call in method dispatch to 'identifier' (no default method)\", \n",
       "        domain = NA))(GdObject, ...))\n",
       "```"
      ],
      "text/plain": [
       "new(\"standardGeneric\", .Data = function (GdObject, ...) \n",
       "standardGeneric(\"identifier\"), generic = \"identifier\", package = \"Gviz\", \n",
       "    group = list(), valueClass = character(0), signature = \"GdObject\", \n",
       "    default = NULL, skeleton = (function (GdObject, ...) \n",
       "    stop(\"invalid call in method dispatch to 'identifier' (no default method)\", \n",
       "        domain = NA))(GdObject, ...))\n",
       "<bytecode: 0x555c884967e8>\n",
       "<environment: 0x555c86518900>\n",
       "attr(,\"generic\")\n",
       "[1] \"identifier\"\n",
       "attr(,\"generic\")attr(,\"package\")\n",
       "[1] \"Gviz\"\n",
       "attr(,\"package\")\n",
       "[1] \"Gviz\"\n",
       "attr(,\"group\")\n",
       "list()\n",
       "attr(,\"valueClass\")\n",
       "character(0)\n",
       "attr(,\"signature\")\n",
       "[1] \"GdObject\"\n",
       "attr(,\"default\")\n",
       "`\\001NULL\\001`\n",
       "attr(,\"skeleton\")\n",
       "(function (GdObject, ...) \n",
       "stop(\"invalid call in method dispatch to 'identifier' (no default method)\", \n",
       "    domain = NA))(GdObject, ...)\n",
       "attr(,\"class\")\n",
       "[1] \"standardGeneric\"\n",
       "attr(,\"class\")attr(,\"package\")\n",
       "[1] \"methods\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "identifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "19b42036",
   "metadata": {},
   "outputs": [],
   "source": [
    "?identifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f7184196",
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in (function (classes, fdef, mtable) : unable to find an inherited method for function 'identifier' for signature '\"DataTrack\"'\n",
     "output_type": "error",
     "traceback": [
      "Error in (function (classes, fdef, mtable) : unable to find an inherited method for function 'identifier' for signature '\"DataTrack\"'\nTraceback:\n",
      "1. identifier(diff.track)",
      "2. (function (classes, fdef, mtable) \n . {\n .     methods <- .findInheritedMethods(classes, fdef, mtable)\n .     if (length(methods) == 1L) \n .         return(methods[[1L]])\n .     else if (length(methods) == 0L) {\n .         cnames <- paste0(\"\\\"\", vapply(classes, as.character, \n .             \"\"), \"\\\"\", collapse = \", \")\n .         stop(gettextf(\"unable to find an inherited method for function %s for signature %s\", \n .             sQuote(fdef@generic), sQuote(cnames)), domain = NA)\n .     }\n .     else stop(\"Internal error in finding inherited methods; didn't return a unique method\", \n .         domain = NA)\n . })(list(structure(\"DataTrack\", package = \"Gviz\")), new(\"standardGeneric\", \n .     .Data = function (GdObject, ...) \n .     standardGeneric(\"identifier\"), generic = structure(\"identifier\", package = \"Gviz\"), \n .     package = \"Gviz\", group = list(), valueClass = character(0), \n .     signature = \"GdObject\", default = NULL, skeleton = (function (GdObject, \n .         ...) \n .     stop(\"invalid call in method dispatch to 'identifier' (no default method)\", \n .         domain = NA))(GdObject, ...)), <environment>)",
      "3. stop(gettextf(\"unable to find an inherited method for function %s for signature %s\", \n .     sQuote(fdef@generic), sQuote(cnames)), domain = NA)"
     ]
    }
   ],
   "source": [
    "GenomicRanges::"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "846fef16",
   "metadata": {},
   "source": [
    "# Run Lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50c34204",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Pull mean difference data from trackj\n",
    "mean.diffs <- as.vector(diff.track@data)\n",
    "mean.diffs <- mean.diffs[!is.na(mean.diffs) & mean.diffs != 0]\n",
    "\n",
    "# Run length embedding tells about the runs themselves\n",
    "embedded <- rle(sign(mean.diffs))\n",
    "tmp.df <- data.frame(sign = as.factor(embedded$values), run_length = embedded$lengths * embedded$values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "570816f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "longest.run <- max(embedded$lengths)\n",
    "median.run <- median(embedded$lengths)\n",
    "\n",
    "LOAD.greater <- sum(mean.diffs > 0)\n",
    "CONT.greater <- sum(mean.diffs < 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35a97633",
   "metadata": {},
   "source": [
    "What is the probability of this?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fb063bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "p <- tmp.df %>%\n",
    "    ggplot(aes(x = run_length)) +\n",
    "    geom_freqpoly() +\n",
    "    xlab(\"Run length\") +\n",
    "    scale_y_log10() +\n",
    "    xlim(c(-1*longest.run, longest.run)) +\n",
    "    wiscR::light_theme() +\n",
    "    geom_vline(xintercept = 0, color = \"red\")\n",
    "\n",
    "p\n",
    "wiscR::save_plot(p, ofile = file.path(idir, \"figs/run-lengths.png\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "55463218",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.sample <- sample(c(-1, 1), replace=TRUE, 1e6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd3d2f26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run length embedding tells about the runs themselves\n",
    "embedded <- rle(sign(random.sample))\n",
    "tmp.df <- data.frame(sign = as.factor(embedded$values), run_length = embedded$lengths * embedded$values)\n",
    "\n",
    "tmp.df %>%\n",
    "    ggplot(aes(x = run_length)) +\n",
    "    geom_freqpoly() +\n",
    "    xlab(\"Run length\") +\n",
    "    scale_y_log10() +\n",
    "    xlim(c(-1*longest.run, longest.run)) +\n",
    "    wiscR::light_theme() +\n",
    "    geom_vline(xintercept = 0, color = \"red\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ced345ef",
   "metadata": {},
   "source": [
    "Since it's `LOAD - Control`, we've got roughly 850,000 positions where LOAD methylation is greater, and 680,000 positions where Control methylation is greater. Is this percentage consistent across chromosomes?"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "4ea9fc3c3a8df0489cd5c25a6a21fb2fa46d2dcfb175338d3042d921f3774f42"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
